gtdynamics/factors/LinkPoseFactor.h

```diff
@@ -0,0 +1,109 @@
+/* ----------------------------------------------------------------------------
+ * GTDynamics Copyright 2020, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * See LICENSE for the license information
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file  LinkPoseFactor.h
+ * @brief Factor relating a parent link and a child link given joint
+ * measurement.
+ * @author Varun Agrawal
+ */
+
+#pragma once
+#include <gtsam/nonlinear/NonlinearFactor.h>
+
+#include "gtdynamics/universal_robot/Joint.h"
+#include "gtdynamics/universal_robot/Link.h"
+
+namespace gtdynamics {
+
+/**
+ * @brief A 2-way factor to relate the parent and child links of a joint given
+ * the joint coordinate as a measurement.
+ *
+ * @tparam JOINT The type of joint which is being constrained.
+ */
+template <typename JOINT>
+class LinkPoseFactor
```

Comment:
Not a big fan of this name since it doesn't really describe what the difference between this and `PoseFactor` is.

I appreciate that it's difficult to come up with a name, and that not liking it without providing an alternative is annoying.  Maybe `PoseFactorMeasurement` or `NumericalPoseFactor` or `PoseFactorWithAngle` something?  Ehh not sure, but I'm not a huge fan of LinkPoseFactor.

PS I'm not opposed to renaming both to `LinkPoseFactor`, but atm I don't see how `Link` describes the difference between these two factors.

Reply:
Yeah that's a good point. I have not been a fan of the `PoseFactor` name either but didn't know what would be a good way to name this and convey intent. Maybe `JointCoordinateFactor` or `JointMeasurementFactor`?

Reply:
Personally I would vote for `JointMeasurementFactor`.

Although it kind of sounds like a prior factor on a joint angle variable, but since that's more like FKprior or just PriorFactor then I think it's sufficiently unambiguous.

Reply:
It's not a prior right? It's a constraint on the link poses based on the joint measurement.

Reply:
Right - so my wording as confusing but what I was trying to say is that it sounds as if `JointMeasurementFactor` is a prior factor, but it is _not_ a prior factor: therefore it could be confusing.  Specifically, `JointMeasurementFactor` sounds like it's a prior factor on the joint angle: `gtsam::PriorFactor<double>(JointAngleKey(id, t), measurement)`

However, I think that because there's more obvious names for prior factors (e.g. FKprior, or just PriorFactor), then it's sufficiently clear that `JointMeasurementFactor` is not a prior factor.

Anyway, we're on the same page so we're good.